stdlib/seq.mli

```diff
@@ -13,97 +13,807 @@
+    [compare cmp xs ys] compares the sequences [xs] and [ys]
+    according to the lexicographic preorder.
+
+    For more details on comparison functions, see {!Array.sort}.
+
+    At least one of the sequences [xs] and [ys] must be finite.
+
+    @since 4.14 *)
+
+(** {1 Constructing sequences} *)
+
+(** The functions in this section are lazy: that is, they return sequences
+    whose elements are computed only when demanded. *)
 
 val empty : 'a t
-(** The empty sequence, containing no elements. *)
+(** [empty] is the empty sequence.
+    It has no elements. Its length is 0. *)
 
 val return : 'a -> 'a t
-(** The singleton sequence containing only the given element. *)
+(** [return x] is the sequence whose sole element is [x].
+    Its length is 1. *)
 
 val cons : 'a -> 'a t -> 'a t
-(** [cons x xs] is the sequence containing the element [x] followed by
-    the sequence [xs] @since 4.11 *)
+(** [cons x xs] is the sequence that begins with the element [x],
+    followed with the sequence [xs].
+
+    Writing [cons (f()) xs] causes the function call [f()]
+    to take place immediately. For this call to be delayed until the
+    sequence is queried, one must instead write
+    [(fun () -> Cons(f(), xs))].
 
-val append : 'a t -> 'a t -> 'a t
-(** [append xs ys] is the sequence [xs] followed by the sequence [ys]
     @since 4.11 *)
 
+val init : int -> (int -> 'a) -> 'a t
+(** [init n f] is the sequence [f 0; f 1; ...; f (n-1)].
+
+    [n] must be nonnegative.
+
+    [init n f] is equivalent to [map f (up 0 n)].
+
+    @raise Invalid_argument if [n] is negative.
+    @since 4.14 *)
+
+val unfold : ('b -> ('a * 'b) option) -> 'b -> 'a t
+(** [unfold] constructs a sequence
+    out of a step function and an initial state.
+
+    If [f u] is [None] then
+    [unfold f u] is the empty sequence.
+    If [f u] is [Some (x, u')] then
+    [unfold f u] is the nonempty sequence [cons x (unfold f u')].
+
+    For example, [unfold (function [] -> None | h :: t -> Some (h, t)) l]
+    is equivalent to [List.to_seq l].
+
+    @since 4.11 *)
+
+val repeat : 'a -> 'a t
+(** [repeat x] is the infinite sequence
+    where the element [x] is repeated indefinitely.
+
+    [repeat x] is equivalent to [cycle (return x)].
+
+    @since 4.14 *)
+
+val forever : (unit -> 'a) -> 'a t
+(** [forever f] is an infinite sequence where every element is produced
+    (on demand) by the function call [f()].
+
+    For instance,
+    [forever Random.bool] is an infinite sequence of random bits.
+
+    [forever f] is equivalent to [map f (repeat ())].
+
+    @since 4.14 *)
+
+val cycle : 'a t -> 'a t
+(** [cycle xs] is the infinite sequence which consists of an infinite
+    number of repetitions of the sequence [xs].
+
+    The sequence [xs] must be nonempty.
+
+    Consuming (a prefix of) the sequence [cycle xs] once
+    can cause the sequence [xs] to be consumed more than once.
+    Therefore, [xs] must be persistent.
+
+    @raise Invalid_argument if [xs] is empty.
```

Comment:
What about having `cycle empty == empty`, instead?

This seems somewhat natural to me, if we consider `cycle x` to be the least fixed point of `append x`.